kaiax/gov/param.go

```diff
@@ -0,0 +1,442 @@
+package gov
+
+import (
+	"bytes"
+	"math/big"
+	"strconv"
+	"strings"
+
+	"github.com/kaiachain/kaia/common"
+)
+
+type canonicalizerT func(v interface{}) (interface{}, error)
+
+type Param struct {
+	Name              string
+	ParamSetFieldName string
+	Canonicalizer     canonicalizerT
+	FormatChecker     func(cv interface{}) bool // validation on canonical value.
+
+	DefaultValue  interface{}
+	VoteForbidden bool
+}
+
+var (
+	addressCanonicalizer canonicalizerT = func(v interface{}) (interface{}, error) {
+		switch v := v.(type) {
+		case []byte:
+			if len(v) != common.AddressLength {
+				return nil, ErrCanonicalizeByteToAddress
+			}
+			return common.BytesToAddress(v), nil
+		case string:
+			if !common.IsHexAddress(v) {
+				return nil, ErrCanonicalizeStringToAddress
+			}
+			return common.HexToAddress(v), nil
+		case common.Address:
+			return v, nil
+		}
+		return nil, ErrCanonicalizeToAddress
+	}
+
+	addressListCanonicalizer canonicalizerT = func(v interface{}) (interface{}, error) {
+		stringToAddressList := func(v string) ([]common.Address, error) {
+			ret := []common.Address{}
+			for _, address := range strings.Split(v, ",") {
+				if !common.IsHexAddress(address) {
+					return nil, ErrCanonicalizeStringToAddress
+				}
+				ret = append(ret, common.HexToAddress(address))
+			}
+			return ret, nil
+		}
+
+		switch v := v.(type) {
+		case []byte:
+			return stringToAddressList(string(v))
+		case string:
+			return stringToAddressList(v)
+		}
+		return nil, ErrCanonicalizeToAddressList
+	}
+
+	bigIntCanonicalizer canonicalizerT = func(v interface{}) (interface{}, error) {
+		switch v := v.(type) {
+		case []byte:
+			cv, ok := new(big.Int).SetString(string(v), 10)
+			if !ok {
+				return nil, ErrCanonicalizeByteToBigInt
+			}
+			return cv, nil
+		case string:
+			cv, ok := new(big.Int).SetString(v, 10)
+			if !ok {
+				return nil, ErrCanonicalizeStringToBigInt
+			}
+			return cv, nil
+		case *big.Int:
+			return v, nil
+		}
+		return nil, ErrCanonicalizeBigInt
+	}
+
+	boolCanonicalizer canonicalizerT = func(v interface{}) (interface{}, error) {
+		switch v := v.(type) {
+		case []byte:
+			if bytes.Equal(v, []byte{0x01}) {
+				return true, nil
+			} else if bytes.Equal(v, []byte{0x00}) {
+				return false, nil
+			} else {
+				return nil, ErrCanonicalizeByteToBool
+			}
+		case bool:
+			return v, nil
+		}
+		return nil, ErrCanonicalizeBool
+	}
+
+	stringCanonicalizer canonicalizerT = func(v interface{}) (interface{}, error) {
+		switch v := v.(type) {
+		case []byte:
+			return string(v), nil
+		case string:
+			return v, nil
+		}
+		return nil, ErrCanonicalizeString
+	}
+
+	uint64Canonicalizer canonicalizerT = func(v interface{}) (interface{}, error) {
```

Comment:
How about adopting generic to restrict allowed parameters  such as:

```go
package main

import (
    "errors"
    "math/big"
)

// Define custom error messages
var (
    ErrCanonicalizeByteToUint64  = errors.New("cannot canonicalize []byte to uint64")
    ErrCanonicalizeFloatToUint64 = errors.New("cannot canonicalize float64 to uint64")
    ErrCanonicalizeInvalidType   = errors.New("invalid type for canonicalization")
)

// Type constraint that allows only []byte, float64, and uint64
type Uint64Canonicalizable interface {
    ~[]byte | ~float64 | ~uint64
}

// The generic canonicalizer function
func uint64Canonicalizer[T Uint64Canonicalizable](v T) (uint64, error) {
    switch val := any(v).(type) {
    case []byte:
        if len(val) > 8 {
            return 0, ErrCanonicalizeByteToUint64
        }
        return new(big.Int).SetBytes(val).Uint64(), nil
    case float64:
        if float64(uint64(val)) != val {
            return 0, ErrCanonicalizeFloatToUint64
        }
        return uint64(val), nil
    case uint64:
        return val, nil
    default:
        return 0, ErrCanonicalizeInvalidType
    }
}

func main() {
    // Examples of usage:
    bytesVal := []byte{0x01, 0x02}
    floatVal := 12345.0
    uintVal := uint64(67890)
    uintVal32 := uint32(67890)

    // Handling []byte
    res1, err1 := uint64Canonicalizer(bytesVal)
    if err1 != nil {
        panic(err1)
    }
    println(res1)

    // Handling float64
    res2, err2 := uint64Canonicalizer(floatVal)
    if err2 != nil {
        panic(err2)
    }
    println(res2)

    // Handling uint64
    res3, err3 := uint64Canonicalizer(uintVal)
    if err3 != nil {
        panic(err3)
    }
    println(res3)

    // Compile time error
    uint64Canonicalizer(uintVal32)   
}
```

Reply:
Sounds great, but then how do I define `canonicalizerT`? 

```go
type Param struct {
	Name              string
	ParamSetFieldName string
	Canonicalizer     canonicalizerT
	FormatChecker     func(cv any) bool // validation on canonical value.

	DefaultValue  any
	VoteForbidden bool
}
```

Reply:
Yes, there's a type of obstacle.. Let it go with the current one.